### PR TITLE
Prevent non course participants from interacting with components.

### LINKED ISSUE #1226: Permission issues
- [ ] User not enrolled in course can see leaderboard
- [ ] User not enrolled in course can see forum
- [x] Non-admin user cannot reply to forum post by clicking the reply button

### PR DESCRIPTION
**Pull request by fonglh**
Fixes #1226 and #1230.

### CODE REVIEW

**Comment by allenwq** [comment]
Add a controller test for forum or assessment to verify ?

**Comment by fonglh** [comment]
@allenwq spec for forum added

File: spec/controllers/course/forum/forums_controller_spec.rb [hunk 1]
```diff
@@ -18,6 +18,16 @@
 
     before { sign_in(user) }
 
+    context 'with a non course user' do
```
Comment (allenwq) [inline spec/controllers/course/forum/forums_controller_spec.rb]: Let's keep this consistent with other specs and use `when`: 'when current user is not enrolled in the course'

**Comment by allenwq** [comment]
@fonglh Thanks !

Actually, further think about it, I lean to merge `read` and `participate` because they are kind of duplicate.

`check_user_participation` is a subset of `load_and_authorize_resource :course` and most of the time the do the same thing, but it needs one more query for each request. 

It might be OK since rails does some caching for us, but it's better if we can merge them.

**Pull request merged**